integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java

```diff
@@ -0,0 +1,484 @@
+         if (b == null) {
+            throw new Exception(connectorName + ": queue " + queueName + " not found");
+         }
+         this.queue = (Queue) b.getBindable();
+
+         this.kafkaProducer = kafkaProducerFactory.create(kafkaConfig(configuration));
+
+         List<PartitionInfo> topicPartitions = kafkaProducer.partitionsFor(topicName);
+         if (topicPartitions == null || topicPartitions.size() == 0) {
+            throw new Exception(connectorName + ": topic " + topicName + " not found");
+         }
+
+         this.retryCount.set(0);
+         this.isStarted.set(true);
+         connect();
+         ActiveMQKafkaLogger.LOGGER.bridgeStarted(connectorName);
+      }
+   }
+
+   public void connect() throws Exception {
+      synchronized (this) {
+         if (!isConnected && isStarted.get()) {
+            isConnected = true;
+            this.queue.addConsumer(this);
+            this.queue.deliverAsync();
+            ActiveMQKafkaLogger.LOGGER.bridgeConnected(connectorName);
+         }
+      }
+   }
+
+   @Override
+   public void disconnect() {
+      synchronized (this) {
+         if (isConnected) {
+            if (queue != null) {
+               this.queue.removeConsumer(this);
+            }
+
+            cancelRefs();
+            if (queue != null) {
+               queue.deliverAsync();
+            }
+            isConnected = false;
+            ActiveMQKafkaLogger.LOGGER.bridgeDisconnected(connectorName);
+         }
+      }
+   }
+
+   @Override
+   public void stop() {
+      synchronized (this) {
+         if (!this.isStarted.get()) {
+            return;
+         }
+         ActiveMQKafkaLogger.LOGGER.bridgeReceivedStopRequest(connectorName);
+
+         disconnect();
+
+         kafkaProducer.close();
+         kafkaProducer = null;
+         this.isStarted.set(false);
+         ActiveMQKafkaLogger.LOGGER.bridgeStopped(connectorName);
+      }
+   }
+
+   @Override
+   public boolean isStarted() {
+      return this.isStarted.get();
+   }
+
+   @Override
+   public String getName() {
+      return this.connectorName;
+   }
+
+   @Override
+   public boolean supportsDirectDelivery() {
+      return false;
+   }
+
+   @Override
+   public HandleStatus handle(MessageReference ref) throws Exception {
+      if (filter != null && !filter.match(ref.getMessage())) {
+         return HandleStatus.NO_MATCH;
+      }
+
+      synchronized (this) {
+         ref.handled();
+
+         Message message = ref.getMessage();
+
+         synchronized (refs) {
+            refs.put(ref.getMessage().getMessageID(), ref);
+         }
+
+         Integer partition = null;
+         SimpleString groupdID = message.getGroupID();
+         if (groupdID != null) {
+            List partitions = kafkaProducer.partitionsFor(topicName);
+            int numPartitions = partitions.size();
+            partition = Utils.toPositive(Utils.murmur2(groupdID.getData())) % numPartitions;
```

Comment:
It's the same way as the DefaultPartitioner works in Kafka, why do you re-used the logic here ?

Reply:
You don't have access to native headers in the partitioner. As such where serializers encode the payload into the value and map message headers (groupID) to kafka native headers, it is not available within kafka partitioner class.

Reply:
"You don't have access to native headers in the partitioner" ...  I didn't get this point. Serialization and Partitioning are two different steps in the Kafka producer pipeline (first serialization, then partitioning). About the partitioner, if you pass the key to the producer send method, it applies exactly this : 

`Utils.toPositive(Utils.murmur2(keyBytes)) % numPartitions`

where `keyBytes` is just  `groupid`. Isn't it ?

Reply:
No keybytes is the LVQ property if set, as key is used for compacted topics.

Reply:
I have an idea of handling this with a custom partitioner and custom serialiser as an alternative

Reply:
Updated so this logic is done with a partitioner (though it is more or less similar still) as we need to use groupID if present to partition by else we fall back to default of using key. But key must be LVQ for compaction topic reasons.